src/plugins/enzymes/src/EnzymesIO.cpp

```diff
@@ -72,25 +72,51 @@ QList<SEnzymeData> EnzymesIO::readEnzymes(const QString& url, U2OpStatus& os) {
     }
 
     QList<SEnzymeData> resToDelete;
-    // assign alphabet if needed.
+    // do additional calculations to recognize some enzymes features
     for (int i = 0, n = res.count(); i < n; i++) {
         SEnzymeData& d = res[i];
         if (d->seq == QByteArray("?")) {
             algoLog.trace(tr("The enzyme '%1' has unknown sequence").arg(d->id));
             resToDelete.append(d);
         } else {
+            // calculate enzyme alphabet
+            d->alphabet = U2AlphabetUtils::findBestAlphabet(d->seq);
             if (d->alphabet == nullptr) {
-                d->alphabet = U2AlphabetUtils::findBestAlphabet(d->seq);
-                if (!d->alphabet) {
-                    algoLog.error(tr("No enzyme alphabet: '%1', sequence '%2'")
-                                      .arg(d->id)
-                                      .arg(QString(d->seq)));
+                algoLog.error(tr("No enzyme alphabet: '%1', sequence '%2'")
+                                    .arg(d->id)
+                                    .arg(QString(d->seq)));
+            } else if (!d->alphabet->isNucleic()) {
+                algoLog.error(tr("Non-nucleic enzyme alphabet: '%1', alphabet: %2, sequence '%3'")
+                                    .arg(d->id)
+                                    .arg(d->alphabet->getId())
+                                    .arg(QString(d->seq)));
+            }
+            // calculate enzyme overhang types
+            auto seqSize = d->seq.size();
+            if (d->cutDirect == ENZYME_CUT_UNKNOWN) {
+                d->overhangTypes |= EnzymeData::OverhangType::NoOverhang;
+            } else {
+                if (qFuzzyCompare((double)seqSize / 2, d->cutDirect) && d->cutDirect == d->cutComplement) {
+                    d->overhangTypes |= EnzymeData::OverhangType::Blunt;
+                } else if ((0 <= d->cutDirect && d->cutDirect < seqSize) &&
+                    (0 <= d->cutComplement && d->cutComplement < seqSize)) {
+                    auto first = d->cutDirect;
+                    auto second = seqSize - d->cutComplement;
+                    auto overhang = d->seq.mid(qMin(first, second), qAbs(first - second));
+                    auto overhangAlphabet = U2AlphabetUtils::findBestAlphabet(overhang);
```

Comment:
It could be `BaseDNAAlphabetIds::NUCL_DNA_DEFAULT()` or `BaseDNAAlphabetIds::NUCL_DNA_EXTENDED()`. The alphabet of overhang could be different from the whole enzyme alphabet.

Reply:
I mean if the algorithm of the whole sequence is NUCL_DNA_DEFAULT any sub-sequence will also have NUCL_DNA_DEFAULT - no need to run alphabet detection for the sub-sequence again in this case.

Reply:
OK